### PR TITLE
Training with DPSGD Implementation

### PR DESCRIPTION
**Pull request by hardlyhuman**
This pull request introduces support for training models using the DPSGD (Differentially Private Stochastic Gradient Descent) algorithm, ensuring differential privacy during training. DPSGD allows the model to learn while preserving the privacy of individual training examples, making it suitable for sensitive data scenarios.

### CODE REVIEW

**Comment by tremblerz** [comment]
1. This looks good but I guess you should use a library (or write on your own) for privacy accounting or ideally pass epsilon as a parameter in the config.
2. Also, can you update the config and put "if/else" wherever necessary so that anyone can use the dpsgd routine.

File: src/algos/base_class.py [hunk 1]
```diff
@@ -575,9 +589,46 @@ def local_train(self, round: int, epochs: int = 1, **kwargs: Any) -> Tuple[float
...
             for _ in range(epochs):
-                tr_loss, tr_acc = self.model_utils.train(
-                    self.model, self.optim, self.dloader, self.loss_fn, self.device, malicious_type=self.config.get("malicious_type", "normal"), config=self.config,
-                )            
+                if self.config.get("use_dpsgd"):
+                    self.log_utils.log_console(
+                        "Training with DPSGD"
+                    )
+                    epsilon = self.config.get("epsilon")
+                    if epsilon:
```
Comment (tremblerz) [inline src/algos/base_class.py:597]: I think this implementation will give incorrect privacy guarantee because for every round we will create a new privacy engine which leads to underestimation of epsilon.

File: src/configs/algo_config.py [hunk 2]
```diff
@@ -32,11 +32,14 @@ def get_malicious_types(malicious_config_list: List[ConfigType]) -> Dict[str, st
...
     "algo": "fedavg",
     "rounds": 2,
+    "use_dpsgd": True,
+    "epsilon": 5,
 
     # Model parameters
     "model": "resnet10",
     "model_lr": 3e-4,
     "batch_size": 256,
+    "epochs": 5,
```
Comment (tremblerz) [inline src/configs/algo_config.py:42]: why do we need `epochs`?

File: src/configs/sys_config.py [hunk 3]
```diff
@@ -316,7 +316,7 @@ def get_digit_five_support(num_users: int, domains: List[str] = DIGIT_FIVE):
     "exp_keys": [],
 }
 
-num_users = 9
+num_users = 4
```
Comment (tremblerz) [inline src/configs/sys_config.py:319]: Please don't make changes to the config other than the ones needed for the feature itself

File: src/utils/model_utils.py [hunk 4]
```diff
@@ -373,6 +373,40 @@ def deep_mutual_train(
             train_accuracy[i] = 100.0 * correct[i] / len(dloader.dataset)
         return train_loss, train_accuracy
 
+
+    def train_with_dpsgd(
```
Comment (tremblerz) [inline src/utils/model_utils.py:377]: Is it any different from the original train method? If not then it would be much more ideal to avoid code duplication

**Review by tremblerz — CHANGES_REQUESTED** [review]

**Comment by tremblerz** [comment]
Inactive by the author so closing it. We will probably do it sometime soon using Opacus
